### ISSUE TITLE
Parser is too large/bloated

### ISSUE THREAD

**Issue by calebdw** [EmranMR/tree-sitter-blade#56]
Hello!

You can determine the size of the parser by looking at the STATE_COUNT in `parser.c`.  The current state count for this parser is 6740 which is extremely large. For comparison, the php parser state count is ~2700, and the phpdoc parser state count is ~339.

The likely culprit here is the `_text` regex which results in a ton of fragments.

**Comment by EmranMR** [comment]
yea, I knew it was very likely the performance getting a hit , because of the RegEx use, but didn't know how to benchmark or analyse! It is always good to have some number to compare with. 

I am really not sure what about any alternatives tbh, any ideas? I believe you were working on a blade parser yourself with scanners, any luck there? That `_text` node and parsing `text` in general has been a nightmare that keeps haunting! #55 !

**Comment by calebdw** [comment]
Yeah I have some ideas that I need to play around with including:

- rewriting the rules
- extending from php or html (or even both)
- external scanner

**Comment by EmranMR** [comment]
I am not sure how extending alone would work in terms of `text`. I suppose that + external scanner might work in harmony?

I actually got the `_text` inspiration from [tree-sitter-html](https://github.com/tree-sitter/tree-sitter-html/blob/949b78051835564bca937565241e5e337d838502/grammar.js#L141), but had to extensively modify it. 

If there is anything at the bottom of the `grammar.js` that you need clarified, please feel free to let me know and I will try to elaborate a bit more. 

I tried to document it as much as I could using comments but yea, it's some ugly code, esp parenthesis balancing recursion bit!

**Comment by calebdw** [comment]
Yes, but interestingly enough the html state count is only 94 bytes

**Comment by EmranMR** [comment]
So based on my tests, the state count hovers around 5.5k. previously any new directive/feature addition was leading to exponential growth. However this is now rectified by using `NodeMap` data structure . 

@calebdw Shall we close this one? 🤔

The grammar is very complex, and Swift being at 6k ish and cpp at 11k I think it is quite reasonable. 

I think the only way to get it super low is via `external_scanner` as you suggested!

**Comment by calebdw** [comment]
Lets leave it open for now as motivation to explore the external scanner

**Comment by clason** [comment]
Drive-by note: while parser size matters, the _real_ important stat that affects performance is `LARGE_STATE_COUNT`.

**Comment by EmranMR** [comment]
@clason do you know what's a good rough size for a grammar? I managed to bring it down to 5-6k (stable, even with new features being added). I see other grammars like Swift, C++ etc are way higher, with robust lexical structure;

considering the complexity of blade, being a markup-ish, on top of html + php, with messy, made-up lexical structure?

The actual parse rate though has been increased 5x fold since v0.12.0 to 20k bytes/ms, which is even faster than even fundamental grammars on my local computer such as `tree-sitter-php`.  

This is 5k more than my last comment above.

**Comment by clason** [comment]
I don't think there's a hard and fast rule. Keeping an eye on the `LARGE_STATE_COUNT` and using https://github.com/ribru17/ts_query_ls to `profile` the references queries when making changes can point to rules that should be refactored.

(For reference, this parser is currently at # 35 in the "state count leaderboard" and at # 55 in the "query parsing time leaderboard". That's roughly top ten percent, but far from the worst offenders.)

**Comment by clason** [comment]
You've probably seen this, but https://github.com/tree-sitter/tree-sitter/wiki/Tips-and-Tricks-for-a-grammar-author has some good insight into how to handle state count increases.
